### PR TITLE
VACMS-5376: Creates a filter to populate node link attributes and trigger Linkit.

### LINKED ISSUE #5376: Linkit should not allow entering unknown internal paths
## Background

The [linkit](https://www.drupal.org/project/linkit) module has two pieces of functionality that work together to allow entering stable internal Drupal paths (e.g. `/node/84`) and have them converted to path aliases (e.g. `/pittsburgh-health-care/locations/pittsburgh-va-medical-center-university-drive`) at render time. This is enabled by the JS autocomplete widget inserting metadata in the `a` tag, like so:

`<a data-entity-substitution="canonical" data-entity-type="node" data-entity-uuid="6025ca20-9912-4688-aa2f-508b7a872107" href="/node/16184">`

Occasionally, this does not happen, and the `a` tag is stored with no metadata. This creates broken link errors, because the node path is not converted to an alias, and the node path is unavailable on VA.gov.

## Impact

Editors can unintentionally create broken links that require triage and resolution by CMS help desk.

## How to Reproduce

In general, any manual input will re-create the problem, even for valid paths. For instance:

- type a known node path, e.g. `/node/84`:
Screen Shot 2021-05-18 at 7 19 37 AM

- instead of clicking on "Pittsburgh VA Medical center - University Drive", click above the field. The path will appear identical whether you click there, or on the autocompleted text:
Screen Shot 2021-05-18 at 7 21 30 AM

- However, if you view source, you will see the node path stored without metadata:
Screen Shot 2021-05-18 at 7 22 06 AM

- When clicking on the autocompleted path, it will look like this instead:
Screen Shot 2021-05-18 at 7 23 01 AM

## Implementation

There are at least 3 potential places where this might be resolved:
- In the widget itself, by not allowing the user to click 'Save' until a known internal path or fully qualified external URL has been entered
- during form validation, by triggering a form error when internal paths without metadata are detected
- as an input filter, by resolving internal paths without requiring the metadata that linkit uses (This would not resolve unknown paths, e.g. `/node/999999999999999`)

We should prioritize a solution that is contributable upstream if possible.

## Acceptance Criteria
- [ ] It is not possible to enter an internal path in the linkit widget without it being rendered to its canonical alias

### PR DESCRIPTION
**Pull request by ndouglas**
## Description

Closes #5376.  I tried to err on the side of doing -- and changing -- as little as possible.

## Testing done
Created some links and saved the node to see how they were treated (suggestions welcome!):

The following source was injected in a WYSIWYG box to resemble the screenshot below:

```
<h2>Connect with a care coordinator</h2>

<p><a href="/node/1">Test1</a> = &lt;a href="/node/1"&gt;node1&lt;/a&gt;</p>

<p><a href="/node/abcde">Test2</a> = &lt;a href="/node/abcde"&gt;node2&lt;/a&gt;</p>

<p><a href="/node/27465">Test3</a> = &lt;a href="/node/27465"&gt;node3&lt;/a&gt;</p>

<p><a href="/node/29203420984230984230482">Test4</a> = &lt;a href="/node/29203420984230984230482"&gt;node4&lt;/a&gt;</p>

<p><a data-entity-substitution="canonical" data-entity-type="node" href="/node/27465">Test5</a> = &lt;a href="/node/27465" data-entity-type="node"&gt;node5&lt;/a&gt;</p>

<p><a data-entity-substitution="canonical" data-entity-uuid="465d7e8f-b8e0-4bf2-b5a9-625690c02eaa" href="/node/27465">Test6</a> = &lt;a href="/node/27465" data-entity-uuid="whatever"&gt;node6&lt;/a&gt;</p>

<p><a data-entity-substitution="canonical" data-entity-type="node" data-entity-uuid="465d7e8f-b8e0-4bf2-b5a9-625690c02eaa" href="/node/27465">Test7</a> = &lt;a href="/node/27465" data-entity-type="node"&nbsp;data-entity-uuid="whatever"&gt;node7&lt;/a&gt;</p>
```

The following shows the equivalence between the markup and the link as rendered.

```
Test1 = <a href="/node/1">node1</a> => http://va-gov-cms.lndo.site/node/1 (invalid node)
Test2 = <a href="/node/abcde">node2</a> => http://va-gov-cms.lndo.site/node/abcde (invalid node ID)
Test3 = <a href="/node/27465">node3</a> => http://va-gov-cms.lndo.site/boston-health-care/health-services/women-veteran-care (valid node ID)
Test4 = <a href="/node/29203420984230984230482">node4</a> => http://va-gov-cms.lndo.site/node/29203420984230984230482 (invalid node ID)
Test5 = <a href="/node/27465" data-entity-type="node">node5</a> => http://va-gov-cms.lndo.site/boston-health-care/health-services/women-veteran-care (valid but not containing all the information injected by Linkit)
Test6 = <a href="/node/27465" data-entity-uuid="whatever">node6</a> => http://va-gov-cms.lndo.site/boston-health-care/health-services/women-veteran-care (ditto)
Test7 = <a href="/node/27465" data-entity-type="node" data-entity-uuid="whatever" data-entity-substitution="canonical">node7</a> => http://va-gov-cms.lndo.site/boston-health-care/health-services/women-veteran-care (as injected by widget, and not screwed up by this filter).
```

## Screenshots
Screen Shot 2021-07-21 at 2 50 34 PM

## QA steps

As user _uid_ with _user_role_
1. Do this
1. Then that
1. Then validate Acceptance Criteria from issue
- [ ] This
- [ ] That
- [ ] The other thing

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [x] `Core Application Team`
- [ ] `Product Support Team`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [x] No announcement is needed for this code change. 
  - [x] Merge & carry on unburdened by announcements

### CODE REVIEW

**Comment by ndouglas** [comment]
I ended up doing it this way because I figured it'd be simpler, safer, and less error-prone than a save hook, but feel free to yell at me if you disagree.  It's okay.  I can take it.

**Comment by ndouglas** [comment]
@acrollet I was gonna try to do the bulk publishing check we discussed during standup but it looks like any cached values for `field_wysiwyg['#processed']` or whatever aren't surviving the DB sync/cache clear process.  I don't have to edit anything, bulk or otherwise, it just works by the time I look at it.

So pretty sure that's not going to be an issue.

I also noted that this doesn't fix links of the form `<a href="https://prod.cms.va.gov/node/57">text</a>`.  It's not challenging for me to handle that too, but the code will obv get uglier and links of that form aren't due to the original UX fail, so IMHO it's outside the scope of this issue.  Thoughts?

**Comment by acrollet** [comment]
> @acrollet I was gonna try to do the bulk publishing check we discussed during standup but it looks like any cached values for `field_wysiwyg['#processed']` or whatever aren't surviving the DB sync/cache clear process. I don't have to edit anything, bulk or otherwise, it just works by the time I look at it.
> 
> So pretty sure that's not going to be an issue.

awesome!

> I also noted that this doesn't fix links of the form `<a href="https://prod.cms.va.gov/node/57">text</a>`. It's not challenging for me to handle that too, but the code will obv get uglier and links of that form aren't due to the original UX fail, so IMHO it's outside the scope of this issue. Thoughts?

I agree that this isn't in scope - it's also a relatively infrequent issue.

File: config/sync/filter.format.rich_text_limited.yml [hunk 1]
```diff
@@ -99,8 +101,26 @@ filters:
     id: media_embed
     provider: media
     status: false
-    weight: 100
+    weight: -36
     settings:
       default_view_mode: default
       allowed_media_types: {  }
       allowed_view_modes: {  }
+  blazy_filter:
```
Comment (acrollet) [inline config/sync/filter.format.rich_text_limited.yml:109]: @ndouglas could I have a ELI5 on what this blazy filter is? Thanks!!

Comment (ndouglas) [reply config/sync/filter.format.rich_text_limited.yml:109]: No clue and it's not enabled in the current config or in the PR preview.  I'm not sure why these changes exist -- maybe just one of those things where changing the config unstuck some other changes.  (There were some others, like the revision log message nightmare, but I filtered those files out)

I'll re-sync the DB and re-export the changes, just to make sure I'm not introducing any cruft or fouling something up.

File: docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function process($text, $langcode) {
+    $result = new FilterProcessResult($text);
+    $dom = Html::load($text);
```
Comment (acrollet) [inline docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:71]: I'm probably over-thinking this but I'm wondering if this round-trip through the Html parser/serializer could cause any unintended changes - is this a common pattern for other Filters? (asking out of my ignorance)

Comment (ndouglas) [reply docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:71]: I think it's about as safe as anything can be and fairly normal behavior.

Ideally, I'd have a guard clause up top and be able to short-circuit it completely, but I don't like mixing HTML and regular expressions and if I'm going to load the DOM and do XPath stuff on it to match elements reliably and cleanly, I think the performance overhead is minimal and forgivable.

Comment (ndouglas) [reply docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:71]: well, the most common forms are all gonna be `<a href="/node"` so I don't need regex for that.  And I probably don't need to worry about the other degenerate cases (missing one attribute but not the other, etc).  BRB adding a guard clause

File: config/sync/filter.format.rich_text_limited.yml (hunk 1 above)
Comment (ndouglas) [reply config/sync/filter.format.rich_text_limited.yml:109]: Yeah, did it from scratch and it still has these changes -- guess it's just YAML we have now 🤷‍♂️

File: docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    $xpath = new \DOMXPath($dom);
+    foreach ($xpath->query('//a[starts-with(@href, "/node/") and ( not(@data-entity-type) or not(@data-entity-uuid) )]') as $element) {
+      $path = $element->getAttribute('href');
+      if (preg_match('#node/(\d+)#', $path, $matches)) {
```
Comment (acrollet) [inline docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:78]: It's very arguable whether this is better but I was thinking the preg_match and a nesting level could be avoided by doing something like:

```
$params = Url::fromUri("internal:{$path}")->getRouteParameters();
if ($params['node'] && $node = $this->entityTypeManager()->getStorage('node')->load($params['node'])) {
  $element->setAttribute('data-entity-type', 'node');
  $element->setAttribute('data-entity-substitution', 'canonical');
  $element->setAttribute('data-entity-uuid', $node->uuid());
}
```

take it or leave it :-)

File: docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php (hunk 2 above)
Comment (ndouglas) [reply docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:74]: Added a guard clause and verified that cases 1, 2, and 4 were left unmodified and the others were still corrected appropriately.  (Relative to above issue with config I started with production config and data and enabled the filter, so it wasn't just cached stuff.)

File: config/sync/filter.format.rich_text_limited.yml (hunk 1 above)
Comment (acrollet) [reply config/sync/filter.format.rich_text_limited.yml:109]: we all live in a YAML submarine, a YAML submarine

File: docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php (hunk 3 above)
Comment (ndouglas) [reply docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:78]: More Drupal-ish, but looking at the source of [`Url::fromUri()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Url.php/function/Url%3A%3AfromUri/9.1.x) I'm not sure what we'd be gaining.

Comment (ndouglas) [reply docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:78]: It would make it easier to generalize if we started having problems with people linking to `/term/2423423`, etc.  That had crossed my mind when I started on this but ended up just ignoring it.

Comment (acrollet) [reply docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:78]: > More Drupal-ish, but looking at the source of [`Url::fromUri()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Url.php/function/Url%3A%3AfromUri/9.1.x) I'm not sure what we'd be gaining.

ah yes, the engineering equivalent of shoving the mess into the closet when mom tells you to clean your room.

Comment (acrollet) [reply docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:78]: like I said it's super arguable, go with your gut

Comment (ndouglas) [reply docroot/modules/custom/va_gov_backend/src/Plugin/Filter/NodeLinkEnforcementFilter.php:78]: Eh, I like it but I think I'll leave it out.

**Comment by acrollet** [comment]
Testing this PR with the use case in #5376:

Deliberately added a link to /node/84 without clicking on the linkit result:

Screen Shot 2021-07-22 at 11 58 23 AM

After saving, the metadata is shown correctly on the proofing page source:

Screen Shot 2021-07-22 at 11 59 05 AM

Rebuild of the FE on the tugboat env is in process....

**Comment by acrollet** [comment]
Confirmed that the link is correctly rendered after a FE build:

Screen Shot 2021-07-22 at 12 13 27 PM

**Review by acrollet — APPROVED** [review]
tests pass, QA & CR pass, awesome work @ndouglas !
